### PR TITLE
Build: Fix compiler warnings

### PR DESCRIPTION
**Pull request by nk1506**
Fixing few xlint related warnings. 
examples:

> Note: /home/runner/work/iceberg/iceberg/nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java uses or overrides a deprecated API.
> 
> Note: Recompile with -Xlint:deprecation for details.

> Note: /Users/naveenkumar/Desktop/dremio_work/git/forked/iceberg/gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSFileIOTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

Ref from one of the [build](https://github.com/apache/iceberg/actions/runs/6465826656/job/17552687522)

Before:

Screenshot 2023-10-10 at 5 31 45 PM

After:

Screenshot 2023-10-10 at 5 35 00 PM

### CODE REVIEW

**Review by jbonofre — APPROVED** [review]
LGTM, thanks !

**Review by Fokko — APPROVED** [review]
Nice one @nk1506 thanks for working on this

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java [hunk 1]
```diff
@@ -477,7 +477,7 @@ public void commitTable(
     Branch branch =
         getApi()
             .commitMultipleOperations()
-            .operation(Operation.Put.of(key, newTable, expectedContent))
```
Comment (dimas-b) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java:480]: This change LGTM, but it's a non-trivial change in the Nessie Catalog, certainly not a simple "compiler warning" kind of change... Would you mind moving it into a separate PR for the sake of clarity?

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java [hunk 2]
```diff
@@ -477,7 +477,7 @@ public void commitTable(
...
+            .operation(Operation.Put.of(key, newTable))
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java:480]: This change LGTM, but it's a non-trivial change in the Nessie Catalog, certainly not a simple "compiler warning" kind of change... 

@nk1506 : Would you mind moving it into a separate PR for the sake of clarity?

[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java:480]: +1 for separate PR. 

I think we can even refactor the method to not pass `expectedContent` but just pass contentID as the whole content is unused now.

[the remark this answers is not in the document]
Comment (dimas-b) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java:480]: Please reference https://github.com/projectnessie/nessie/pull/6438 as the rationale for removing the third parameter here.

[the remark this answers is not in the document]
Comment (dimas-b) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java:480]: I believe Nessie API v1 only needs to support older clients to serialize the `expectedContent` parameters in JSON. Actual values are not used in Nessie Servers 0.54.0 and later.

**Review by ajantha-bhat — APPROVED** [review]

File: spark/v3.5/spark/src/jmh/java/org/apache/iceberg/spark/action/IcebergSortCompactionBenchmark.java [hunk 3]
```diff
@@ -280,6 +280,7 @@ protected Configuration initHadoopConf() {
     return new Configuration();
   }
 
+  @SuppressWarnings("unchecked")
```
Comment (nastra) [inline spark/v3.5/spark/src/jmh/java/org/apache/iceberg/spark/action/IcebergSortCompactionBenchmark.java:283]: I'd prefer to fix the underlying issue here rather than just suppressing this.

```
-    SparkSessionCatalog catalog;
+    SparkSessionCatalog<?> catalog;
     try {
       catalog =
-          (SparkSessionCatalog) Spark3Util.catalogAndIdentifier(spark(), "spark_catalog").catalog();
+          (SparkSessionCatalog<?>) Spark3Util.catalogAndIdentifier(spark(), "spark_catalog").catalog();
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 4]
```diff
@@ -119,6 +119,7 @@ public static String translateToIcebergProp(String hmsProp) {
   private final FileIO fileIO;
   private final ClientPool<IMetaStoreClient, TException> metaClients;
 
+  @SuppressWarnings("unchecked")
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:122]: I would rather fix the underlying issue: 
```
-      ClientPool metaClients,
+      ClientPool<IMetaStoreClient, TException> metaClients,
```

**Review by nastra — CHANGES_REQUESTED** [review]
In some cases I believe we can fix the code rather than suppressing the warning
